GpuJsonToStructs fails with token count assertion on multiple malformed open-brace rows

## Summary

This is a minimal reproducer for the same cuDF JSON tokenizer assertion seen in a customer Spark RAPIDS 25.08.0 workload:

```text
nested_json_gpu.cu:1689: Generated token count exceeds the expected token count
```

This issue is distinct from:

- #14811: `GpuGetJsonObject` nesting-depth abort in `get_json_object.cu`
- #14812: `GpuJsonToStructs` `table_view.cpp: Column size mismatch`

This issue specifically reproduces the customer-style `GpuJsonToStructs` failure in `JSONUtils.fromJSONToStructs` with a very small malformed JSON input.

## Environment

Verified on both customer logs and dev-box repros:

- Spark: 3.5.1 / 3.5.3
- RAPIDS Accelerator: 25.08.0
- spark-rapids tag: `v25.08.0` / `43dd3ae8c8bf91ec51e31e2a438417f2951724bd`
- spark-rapids-jni tag: `v25.08.0` / `155ef36a7e5c38e404d76976a61d177354c99281`
- cuDF tag: `v25.08.00` / `6cea3743b6d1da02303e9b3ad819c8dfab77dfc5`
- CUDA build: cuda12 jar

I also checked with RAPIDS 26.04.2 for related JSON failures; the broader JSON failures remain reproducible there as well, but this issue was minimized on 25.08.0.

## Minimal Reproducer

The minimal failing input is two rows containing only an open brace:

```text
{
{
```

Parsing one row succeeds/recover-null, but parsing two rows in the same batch fails with the token count assertion.

Relevant PySpark reproducer:

```python
from pyspark.sql import SparkSession
from pyspark.sql.functions import col, from_json
from pyspark.sql.types import StringType, StructField, StructType

spark = SparkSession.builder.appName("repro-gpu-json-token-count").getOrCreate()
spark.sparkContext.setLogLevel("WARN")

schema = StructType([StructField("a", StringType(), True)])
input_dir = "/tmp/repro_gpu_json_token_count/input"
output_dir = "/tmp/repro_gpu_json_token_count/output"

data = [(0, "{"), (1, "{")]
spark.createDataFrame(data, ["id", "json"]).coalesce(1).write.mode("overwrite").parquet(input_dir)

df = spark.read.parquet(input_dir).coalesce(1)
out = df.select("id", from_json(col("json"), schema).alias("parsed"))
out.write.mode("overwrite").parquet(output_dir)
```

Run command:

```bash
TZ=UTC spark-submit \
  --master local[2] \
  --driver-memory 8g \
  --jars ~/work/rapids.jar \
  --conf spark.plugins=com.nvidia.spark.SQLPlugin \
  --conf spark.rapids.sql.enabled=true \
  --conf spark.rapids.sql.explain=NONE \
  --conf spark.eventLog.enabled=false \
  --conf spark.sql.session.timeZone=UTC \
  --conf spark.driver.extraJavaOptions=-Duser.timezone=UTC \
  --conf spark.executor.extraJavaOptions=-Duser.timezone=UTC \
  --conf spark.sql.shuffle.partitions=1 \
  --conf spark.rapids.sql.batchSizeBytes=256m \
  --conf spark.rapids.sql.concurrentGpuTasks=2 \
  repro_gpu_json_token_count.py
```

I initially found this with a probe script that tested repeated short malformed rows:

```text
CASE_OK open_brace 1 1
CASE_FAILED open_brace 2 1 '{'
```

## Observed Error

```text
org.apache.spark.SparkException: Task failed while writing rows.
  at org.apache.spark.sql.rapids.GpuFileFormatWriterBase.executeTask(GpuFileFormatWriter.scala:445)
  ...
Caused by: ai.rapids.cudf.CudfException: CUDF failure at:
/home/jenkins/agent/workspace/jenkins-spark-rapids-jni-release-24-cuda12/thirdparty/cudf/cpp/src/io/json/nested_json_gpu.cu:1689:
Generated token count exceeds the expected token count
  at com.nvidia.spark.rapids.jni.JSONUtils.fromJSONToStructs(Native Method)
  at com.nvidia.spark.rapids.jni.JSONUtils.fromJSONToStructs(JSONUtils.java:191)
  at org.apache.spark.sql.rapids.GpuJsonToStructs.$anonfun$doColumnar$1(GpuJsonToStructs.scala:56)
  at org.apache.spark.sql.rapids.GpuJsonToStructs.doColumnar(GpuJsonToStructs.scala:51)
```

The corresponding cuDF code in `v25.08.00` computes a conservative token output estimate, then asserts:

```cpp
auto const max_token_out_count =
  cudf::util::div_rounding_up_safe(json_in.size(), min_chars_per_struct) * max_tokens_per_struct;
...
CUDF_EXPECTS(num_total_tokens <= max_token_out_count,
             "Generated token count exceeds the expected token count");
```

For a batch of two malformed one-character JSON strings (`"{"`, `"{"`), the recover/tokenization path appears to generate more tokens than this estimate.

## Expected Behavior

For malformed JSON input, Spark CPU `from_json` recovers by returning null rather than aborting the query. The GPU path should behave consistently, or fall back to CPU if this recovery mode cannot be safely supported.

## Actual Behavior

The GPU path aborts the whole Spark job with a native cuDF token count assertion.

## Customer Context

This reproducer was found while investigating a workload. The customer run failed in the final GPU write path with the same cuDF error:

```text
GpuFileFormatWriter -> GpuProjectExec -> GpuJsonToStructs -> JSONUtils.fromJSONToStructs
nested_json_gpu.cu:1689: Generated token count exceeds the expected token count
```

The minimal two-row `{` reproducer appears to isolate the same underlying `GpuJsonToStructs` / cuDF tokenizer assertion without requiring customer data.

--- a/cpp/src/io/json/nested_json_gpu.cu
+++ b/cpp/src/io/json/nested_json_gpu.cu
@@ -1631,6 +1631,11 @@ std::pair<rmm::device_uvector<PdaTokenT>, rmm::device_uvector<SymbolOffsetT>> ge
       tokenizer_pda::get_translation_table(recover_from_error)),
     stream);
 
+  // In case we're recovering on invalid JSON lines, post-processing the token stream requires to
+  // see a JSON-line delimiter as the very first item
+  SymbolOffsetT const delimiter_offset =
+    (format == tokenizer_pda::json_format_cfg_t::JSON_LINES_RECOVER ? 1 : 0);
+
   // Perform a PDA-transducer pass
   // Compute the maximum amount of tokens that can possibly be emitted for a given input size
   // Worst case ratio of tokens per input char is given for a struct with an empty field name, that
@@ -1639,13 +1644,10 @@ std::pair<rmm::device_uvector<PdaTokenT>, rmm::device_uvector<SymbolOffsetT>> ge
   std::size_t constexpr min_chars_per_struct  = 5;
   std::size_t constexpr max_tokens_per_struct = 6;
   auto const max_token_out_count =
-    cudf::util::div_rounding_up_safe(json_in.size(), min_chars_per_struct) * max_tokens_per_struct;
+    cudf::util::div_rounding_up_safe(json_in.size(), min_chars_per_struct) * max_tokens_per_struct +
+    delimiter_offset;
   cudf::detail::device_scalar<std::size_t> num_written_tokens{
     stream, cudf::get_current_device_resource_ref()};
-  // In case we're recovering on invalid JSON lines, post-processing the token stream requires to
-  // see a JSON-line delimiter as the very first item
-  SymbolOffsetT const delimiter_offset =
-    (format == tokenizer_pda::json_format_cfg_t::JSON_LINES_RECOVER ? 1 : 0);
 
   // Run FST to estimate the size of output buffers
   json_to_tokens_fst.Transduce(zip_in,

--- a/cpp/tests/io/json/json_test.cpp
+++ b/cpp/tests/io/json/json_test.cpp
@@ -2062,11 +2062,6 @@ TEST_F(JsonReaderTest, JSONLinesRecovering)
 
 TEST_F(JsonReaderTest, JSONLinesRecoveringMalformedOpenBraces)
 {
-  /**
-   * @brief Test that recovery mode handles multiple malformed lines with just open braces.
-   * This is a regression test for a bug where the token count estimate didn't account for
-   * the delimiter_offset in recovery mode, causing an assertion failure.
-   */
   std::string data =
     // Two lines with just an open brace (malformed JSON)
     "{\n"

--- a/cpp/tests/io/json/json_test.cpp
+++ b/cpp/tests/io/json/json_test.cpp
@@ -2073,8 +2073,9 @@ TEST_F(JsonReaderTest, JSONLinesRecoveringMalformedOpenBraces)
     "{";
 
   cudf::io::json_reader_options in_options =
-    cudf::io::json_reader_options::builder(cudf::io::source_info{cudf::host_span<std::byte const>{
-      reinterpret_cast<std::byte const*>(data.data()), data.size()}})
+    cudf::io::json_reader_options::builder(
+      cudf::io::source_info{cudf::host_span<std::byte const>{
+        reinterpret_cast<std::byte const*>(data.data()), data.size()}})
       .lines(true)
       .recovery_mode(cudf::io::json_recovery_mode_t::RECOVER_WITH_NULL);
 

--- a/cpp/tests/io/json/json_test.cpp
+++ b/cpp/tests/io/json/json_test.cpp
@@ -2060,6 +2060,30 @@ TEST_F(JsonReaderTest, JSONLinesRecovering)
                     c_validity.cbegin()});
 }
 
+TEST_F(JsonReaderTest, JSONLinesRecoveringMalformedOpenBraces)
+{
+  /**
+   * @brief Test that recovery mode handles multiple malformed lines with just open braces.
+   * This is a regression test for a bug where the token count estimate didn't account for
+   * the delimiter_offset in recovery mode, causing an assertion failure.
+   */
+  std::string data =
+    // Two lines with just an open brace (malformed JSON)
+    "{\n"
+    "{";
+
+  cudf::io::json_reader_options in_options =
+    cudf::io::json_reader_options::builder(cudf::io::source_info{cudf::host_span<std::byte const>{
+      reinterpret_cast<std::byte const*>(data.data()), data.size()}})
+      .lines(true)
+      .recovery_mode(cudf::io::json_recovery_mode_t::RECOVER_WITH_NULL);
+
+  cudf::io::table_with_metadata result = cudf::io::read_json(in_options);
+
+  // All rows are invalid with no schema to infer, so we expect 0 rows
+  EXPECT_EQ(result.tbl->num_rows(), 0);
+}
+
 TEST_F(JsonReaderTest, JSONLinesRecoveringIgnoreExcessChars)
 {
   /**

